File: java/core/joynr-types/src/main/java/io/joynr/pubsub/SubscriptionQos.java
package io.joynr.pubsub;

/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2013 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import io.joynr.subtypes.JoynrType;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public abstract class SubscriptionQos implements JoynrType {
    private static final Logger logger = LoggerFactory.getLogger(SubscriptionQos.class);

    private long expiryDate;
    private long publicationTtl;

    public static final int IGNORE_VALUE = -1;
    public static final long INFINITE_SUBSCRIPTION = Long.MAX_VALUE;
    private static final long MIN_PUBLICATION_TLL = 100L;
    private static final long MAX_PUBLICATION_TLL = 2592000000L; // 30 days

    protected static final long DEFAULT_PUBLICATION_TTL = 10000;

    protected SubscriptionQos() {
    }

    /**
     * 
     * @param expiryDate
     *            The expiryDate is the end date of the subscription. This value is provided in milliseconds (since
     *            1970-01-01T00:00:00.000).
     */
    public SubscriptionQos(long expiryDate) {
        this(expiryDate, DEFAULT_PUBLICATION_TTL);
    }

    /**
     * 
     * @param expiryDate
     *            : the end date of the subscription. This value is provided in milliseconds (since
     *            1970-01-01T00:00:00.000).
     * @param publicationTtl
     *            is the time-to-live for publication messages.<br>
     * <br>
     *            If a notification message can not be delivered within its time to live, it will be deleted from the
     *            system. This value is provided in milliseconds. <br>
     * <br>
     *            <b>Minimum and Maximum Values:</b>
     *            <ul>
     *            <li>minimum publicationTtl_ms = 100. Smaller values will be rounded up.
     *            <li>maximum publicationTtl_ms = 2 592 000 000 (30 days). Larger values will be rounded down.
     *            <li>defualt publicationTtl_ms = 10 000 (10 secs)
     *            </ul>
     */
    public SubscriptionQos(long expiryDate, long publicationTtl) {
        long now = System.currentTimeMillis();
        if (expiryDate <= now) {
            logger.error("Subscription ExpiryDate {} is in the past. Now: {}", expiryDate, now);
        }
        this.expiryDate = expiryDate;
        publicationTtl = publicationTtl < MIN_PUBLICATION_TLL ? MIN_PUBLICATION_TLL : publicationTtl;
        publicationTtl = publicationTtl > MAX_PUBLICATION_TLL ? MAX_PUBLICATION_TLL : publicationTtl;
        this.publicationTtl = publicationTtl;
    }

    /**
     * 
     * @return endDate_ms : the end date of the subscription. <br>
     *         This value is provided in milliseconds (since 1970-01-01T00:00:00.000).
     * 
     */
    public long getExpiryDate() {
        return expiryDate;
    }

    /**
     * 
     * @param expiryDate_ms
     *            is the end date of the subscription. <br>
     *            This value is provided in milliseconds (since 1970-01-01T00:00:00.000).
     * 
     */
    public void setExpiryDate(final long expiryDate_ms) {
        this.expiryDate = expiryDate_ms;
    }

    /**
     * Notification messages will be sent with this time-to-live.<br>
     * <br>
     * If a notification message can not be delivered within its time to live, it will be deleted from the system. This
     * value is provided in milliseconds.
     * 
     * @return publicationTtl_ms time-to-live in milliseconds.
     * 
     */
    public long getPublicationTtl() {
        return publicationTtl;
    }

    /**
     * Notification messages will be sent with this time-to-live. If a notification message can not be delivered within
     * its time to live, it will be deleted from the system. This value is provided in milliseconds.
     * 
     * @param publicationTtl_ms
     *            sets publicationTtl_ms time-to-live in milliseconds.
     *            <ul>
     *            <li>minimum publicationTtl_ms = 100. Smaller values will be rounded up.
     *            <li>maximum publicationTtl_ms = 2 592 000 000 (30 days). Larger values will be rounded down.
     *            </ul>
     * 
     */
    public void setPublicationTtl(final long publicationTtl_ms) {
        this.publicationTtl = publicationTtl_ms;
    }
}

File: cpp/common/SubscriptionQos.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2013 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
#include "joynr/SubscriptionQos.h"
#include "joynr/DispatcherUtils.h"

namespace joynr {

const qint64& SubscriptionQos::DEFAULT_PUBLICATION_TTL() {
    static const qint64 defaultPublicationTtl = 10000;
    return defaultPublicationTtl;
}

const qint64& SubscriptionQos::MIN_PUBLICATION_TTL() {
    static const qint64 minPublicationTtl = 100;
    return minPublicationTtl;
}

const qint64& SubscriptionQos::MAX_PUBLICATION_TTL() {
    static const qint64 maxPublicationTtl = 2592000000UL;
    return maxPublicationTtl;
}

const qint64& SubscriptionQos::NO_EXPIRY_DATE() {
    static qint64 noExpiryDate = 0;
    return noExpiryDate;
}

SubscriptionQos::SubscriptionQos():
    expiryDate(-1),
    publicationTtl(DEFAULT_PUBLICATION_TTL())
{
    setValidity(1000);
}

SubscriptionQos::SubscriptionQos(const qint64& validity) :
    expiryDate(-1),
    publicationTtl(DEFAULT_PUBLICATION_TTL())
{
    setValidity(validity);
}

SubscriptionQos::SubscriptionQos(const SubscriptionQos& subscriptionQos) :
    QObject(),
    expiryDate(subscriptionQos.expiryDate),
    publicationTtl(subscriptionQos.publicationTtl)
{
}

SubscriptionQos::~SubscriptionQos()
{
}

qint64 SubscriptionQos::getPublicationTtl() const {
    return publicationTtl;
}

void SubscriptionQos::setPublicationTtl(const qint64& publicationTtl){
    this->publicationTtl = publicationTtl;
    if(this->publicationTtl > MAX_PUBLICATION_TTL()) {
        this->publicationTtl = MAX_PUBLICATION_TTL();
    }
    if(this->publicationTtl < MIN_PUBLICATION_TTL()) {
        this->publicationTtl = MIN_PUBLICATION_TTL();
    }
}

qint64 SubscriptionQos::getExpiryDate() const {
    return expiryDate;
}

void SubscriptionQos::setExpiryDate(const qint64& expiryDate){
    this->expiryDate = expiryDate;
    if (this->expiryDate < QDateTime::currentMSecsSinceEpoch()) {
        clearExpiryDate();
    }
}

void SubscriptionQos::clearExpiryDate() {
    this->expiryDate = NO_EXPIRY_DATE();
}

void SubscriptionQos::setValidity(const qint64 &validty) {
    setExpiryDate(QDateTime::currentMSecsSinceEpoch() + validty);
}

SubscriptionQos& SubscriptionQos::operator=(const SubscriptionQos& subscriptionQos) {
    expiryDate = subscriptionQos.getExpiryDate();
    publicationTtl = subscriptionQos.getPublicationTtl();
    return *this;
}

bool SubscriptionQos::operator==(const SubscriptionQos& subscriptionQos) const {
    return
        getExpiryDate() == subscriptionQos.getExpiryDate() &&
        publicationTtl == subscriptionQos.getPublicationTtl();
}

bool SubscriptionQos::equals(const QObject &other) const {
    int typeThis = QMetaType::type(this->metaObject()->className());
    int typeOther = QMetaType::type(other.metaObject()->className());
    auto newOther = dynamic_cast<const SubscriptionQos*>(&other);
    return typeThis == typeOther && *this == *newOther;
}
} // namespace joynr
